File: pa01.py
import sys

# Function to read the key file
def read_key(file_name):
    with open(file_name, 'r') as key_file:
        n = int(key_file.readline().strip())  # Read the matrix size
        key_matrix = []
        for _ in range(n):
            row = list(map(int, key_file.readline().strip().split()))
            key_matrix.append(row)  # Read each row of the matrix
    return key_matrix, n

# Function to read and process the plaintext file
def process_plaintext(file_name, n):
    with open(file_name, 'r') as plain_file:
        text = plain_file.read().lower()  # Read all text and convert to lowercase
        filtered_text = ''.join(filter(str.isalpha, text))  # Keep only alphabetic characters
        # Pad with 'x' if the text length is not a multiple of n
        while len(filtered_text) % n != 0:
            filtered_text += 'x'
    return filtered_text

# Function to encrypt the plaintext using the key matrix (Hill cipher)
def hill_cipher_encrypt(plaintext, key_matrix, n):
    ciphertext = []
    for i in range(0, len(plaintext), n):
        block = [ord(char) - ord('a') for char in plaintext[i:i+n]]  # Convert block to numbers
        # Encrypt block by matrix multiplication
        encrypted_block = [(sum(key_matrix[j][k] * block[k] for k in range(n)) % 26) for j in range(n)]
        ciphertext.extend([chr(c + ord('a')) for c in encrypted_block])  # Convert back to characters
    return ''.join(ciphertext)

# Main function
def main():
    if len(sys.argv) != 3:
        print("Usage: python3 pa01.py <key_file> <plaintext_file>")
        sys.exit(1)

    # Read the key file and plaintext file
    key_matrix, n = read_key(sys.argv[1])
    plaintext = process_plaintext(sys.argv[2], n)
    
    # Encrypt the plaintext
    ciphertext = hill_cipher_encrypt(plaintext, key_matrix, n)

    # Output the key matrix
    print("Key matrix:")
    for row in key_matrix:
        print(' '.join(f'{num:4d}' for num in row))

    # Output the processed plaintext
    print("\nPlaintext:")
    for i in range(0, len(plaintext), 80):  # Print 80 characters per line
        print(plaintext[i:i+80])

    # Output the ciphertext
    print("\nCiphertext:")
    for i in range(0, len(ciphertext), 80):  # Print 80 characters per line
        print(ciphertext[i:i+80])

if __name__ == "__main__":
    main()

File: pa01.c
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>

#define MAX_KEY_SIZE 9
#define MAX_TEXT_SIZE 10000

// Function prototypes
void readKeyFile(const char *filename, int *keySize, int key[MAX_KEY_SIZE][MAX_KEY_SIZE]);
void readPlainTextFile(const char *filename, char *plaintext);
void processPlaintext(char *plaintext, char *processedText);
void encryptHill(int keySize, int key[MAX_KEY_SIZE][MAX_KEY_SIZE], char *processedText, char *ciphertext);
void printOutput(int keySize, int key[MAX_KEY_SIZE][MAX_KEY_SIZE], char *plaintext, char *ciphertext);

int main(int argc, char *argv[]) {
    if (argc != 3) {
        fprintf(stderr, "Usage: %s <key_file> <plaintext_file>\n", argv[0]);
        return 1;
    }

    int keySize;
    int key[MAX_KEY_SIZE][MAX_KEY_SIZE];
    char plaintext[MAX_TEXT_SIZE];
    char processedText[MAX_TEXT_SIZE];
    char ciphertext[MAX_TEXT_SIZE];

    readKeyFile(argv[1], &keySize, key);
    readPlainTextFile(argv[2], plaintext);
    processPlaintext(plaintext, processedText);
    encryptHill(keySize, key, processedText, ciphertext);
    printOutput(keySize, key, processedText, ciphertext);

    return 0;
}

void readKeyFile(const char *filename, int *keySize, int key[MAX_KEY_SIZE][MAX_KEY_SIZE]) {
    FILE *file = fopen(filename, "r");
    if (file == NULL) {
        perror("Error opening key file");
        exit(1);
    }

    fscanf(file, "%d", keySize);
    for (int i = 0; i < *keySize; i++) {
        for (int j = 0; j < *keySize; j++) {
            fscanf(file, "%d", &key[i][j]);
        }
    }

    fclose(file);
}

void readPlainTextFile(const char *filename, char *plaintext) {
    FILE *file = fopen(filename, "r");
    if (file == NULL) {
        perror("Error opening plaintext file");
        exit(1);
    }

    int c, i = 0;
    while ((c = fgetc(file)) != EOF && i < MAX_TEXT_SIZE - 1) {
        plaintext[i++] = c;
    }
    plaintext[i] = '\0';

    fclose(file);
}

void processPlaintext(char *plaintext, char *processedText) {
    int i = 0, j = 0;
    while (plaintext[i] != '\0') {
        if (isalpha(plaintext[i])) {
            processedText[j++] = tolower(plaintext[i]);
        }
        i++;
    }
    processedText[j] = '\0';
}

void encryptHill(int keySize, int key[MAX_KEY_SIZE][MAX_KEY_SIZE], char *processedText, char *ciphertext) {
    int textLen = strlen(processedText);
    int paddedLen = ((textLen + keySize - 1) / keySize) * keySize;
    char paddedText[MAX_TEXT_SIZE];
    
    // Pad the text with 'x' if necessary
    strcpy(paddedText, processedText);
    for (int i = textLen; i < paddedLen; i++) {
        paddedText[i] = 'x';
    }
    paddedText[paddedLen] = '\0';

    // Encrypt the padded text
    for (int i = 0; i < paddedLen; i += keySize) {
        int block[MAX_KEY_SIZE] = {0};
        int result[MAX_KEY_SIZE] = {0};

        // Convert a block of letters to numbers (a=0, b=1, ...)
        for (int j = 0; j < keySize; j++) {
            block[j] = paddedText[i + j] - 'a';
        }

        // Perform matrix multiplication
        for (int j = 0; j < keySize; j++) {
            for (int k = 0; k < keySize; k++) {
                result[j] += key[j][k] * block[k];
            }
            result[j] %= 26;  // Modulo 26 to keep within alphabet range
        }

        // Convert numbers back to letters
        for (int j = 0; j < keySize; j++) {
            ciphertext[i + j] = result[j] + 'a';
        }
    }
    ciphertext[paddedLen] = '\0';
}

void printOutput(int keySize, int key[MAX_KEY_SIZE][MAX_KEY_SIZE], char *plaintext, char *ciphertext) {
    printf("Key matrix:\n");
    for (int i = 0; i < keySize; i++) {
        for (int j = 0; j < keySize; j++) {
            printf("%4d", key[i][j]);
        }
        printf("\n");
    }

    printf("\nPlaintext:\n%s\n", plaintext);

    printf("\nCiphertext:\n");
    int len = strlen(ciphertext);
    for (int i = 0; i < len; i++) {
        printf("%c", ciphertext[i]);
        if ((i + 1) % 80 == 0) {
            printf("\n");
        }
    }
    if (len % 80 != 0) {
        printf("\n");
    }
}